homeassistant/components/tuya/fan.py

```diff
@@ -204,6 +205,8 @@ def is_on(self) -> bool | None:
         """Return true if fan is on."""
         if self._switch is None:
             return None
+        if self._speeds is not None and "off" in self._speeds.range:
+            return self.device.status.get(self._speeds.dpcode) != "off"
```

Comment:
This seems to work around an incompatibility, looking at <https://solution.tuya.com/projects/CMatu31ap80406>

It seems like the switch DP, is for controlling the power of the whole device, yet, this entity will at this point use it for controlling the fan.

This entity need to be refactored using EntityDescriptions in order to make that distinction correctly. For this specific device category, I would expect another switch entity for controlling the power (and thus the `switch` DP not being used by this fan entity at all).

Reply:
I've added the EntityDescriptions like you asked and implemented 'fake' on/off behaviour when there is no switch defined for the fan.

I've also made a bunch of changes to exclude the 'off' value from the speed enum because it's not expected by the percentage util (see https://github.com/home-assistant/core/blob/3bb435c292533033c793b5a72aad1a7749c50d7f/homeassistant/util/percentage.py#L12). I think this can be done a bit nicer in places, but since Python is exactly my main language I could use some help there.